packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1244,3 +1254,173 @@ export function validateEventListenerTarget(
+
+export function findFiberRoot(node: Instance): null | FiberRoot {
+  const stack = [node];
+  let index = 0;
+  while (index < stack.length) {
+    const current = stack[index++];
+    if (isContainerMarkedAsRoot(current)) {
+      return ((getInstanceFromNodeDOMTree(current): any): FiberRoot);
+    }
+    stack.push(...current.children);
+  }
+  return null;
+}
+
+export function getBoundingRect(node: Instance): BoundingRect {
+  const rect = node.getBoundingClientRect();
+  return {
+    x: rect.left,
+    y: rect.top,
+    width: rect.width,
+    height: rect.height,
+  };
+}
+
+export function matchAccessibilityRole(node: Instance, role: string): boolean {
+  if (role === getRole(node)) {
+    return true;
+  }
+
+  return false;
+}
+
+export function getTextContent(fiber: Fiber): string | null {
+  switch (fiber.tag) {
+    case HostComponent:
+      let textContent = '';
+      const childNodes = fiber.stateNode.childNodes;
+      for (let i = 0; i < childNodes.length; i++) {
+        const childNode = childNodes[i];
+        if (childNode.nodeType === Node.TEXT_NODE) {
+          textContent += childNode.textContent;
+        }
+      }
+      return textContent;
+    case HostText:
+      return fiber.stateNode.textContent;
+  }
+
+  return null;
+}
+
+export function isHiddenSubtree(fiber: Fiber): boolean {
+  return fiber.tag === HostComponent && fiber.memoizedProps.hidden === true;
+}
+
+export function setFocusIfFocusable(node: Instance): boolean {
+  if (node.nodeType !== Node.ELEMENT_NODE) {
+    // Text and comment nodes aren't focusable.
+    // Technically this check should not be necessary,
+    // since React treats elements (Instances) and text (TextInstance) differently.
+    return false;
+  }
+
+  if (((node: any): HTMLInputElement).disabled === true) {
+    // Disabled inputs can't be focused.
+    return false;
+  }
+
+  const element = ((node: any): HTMLElement);
+
+  if (element.tabIndex === null || element.tabIndex < 0) {
+    // The HTML spec says that negative tab index values indicate an element should be,
+    // "click focusable but not sequentially focusable".
+    // https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute
+    //
+    // The HTML focusable spec also says,
+    // "User agents should consider focusable areas with non-null tabindex values to be click focusable."
+    // https://html.spec.whatwg.org/multipage/interaction.html#focusable
+    //
+    // Despite this, it seems like some browsers (e.g. Chrome, Firefox) return -1 even for elements
+    // that don't accept focus, like HTMLImageElement or the outermost HTMLElement tag.
+    // I think this method should (at least for now) only concern itself with "sequentially focusable" elements.
+    // https://html.spec.whatwg.org/multipage/interaction.html#sequentially-focusable
+    return false;
+  }
+
+  if (element.offsetWidth === 0 || element.offsetHeight === 0) {
+    // Hidden items can't be focused.
```

Comment:
Do you mean they shouldn't be focuseable or they can't? Because they [get focused when calling focus() and are in tab order](https://codesandbox.io/s/tender-breeze-m9k1c?file=/src/App.js).

This would also mean that this wouldn't be usable in JSDOM in which [these values are always 0](https://github.com/jsdom/jsdom/issues/135) because it has no layout engine

Reply:
Shouldn't be. This is simulating what a user would do when clicking on a page.

Reply:
So it's only concerned with clicking? The description sounded like it cares about tab order and these elements are in tab-order.

Reply:
Sorry, they significant part of that statement wasn't the "click" but rather- what a user would do to interact with the page. I was trying to differentiate between _what JavaScript can do_ and _what a user does_. (This was a lightweight attempt at identifying non-displayed stuff.)

I'm not sure about the Code Sandbox you linked. I guess maybe the condition could be changed from `||` to `&&`.

Reply:
> I'm not sure about the Code Sandbox you linked.

You can tab through the page and the button gets focused even if it has `offsetHeight === 0 && offsetWidth === 0`. That's why I'm curious where this condition stems from.

Reply:
I know you can tab through, I saw the sandbox :smile:

The previous version of the Sandbox you linked did not have an `offsetWidth` of 0 for me (which is why I mentioned maybe `&&` instead of `||`).

Reply:
Yeah I added it after you mentioned the extra condition. Wasn't meant to trick you :smile:

Reply:
Closing this loop: I just decided to remove my attempts at detecting if an element is focusable or not and instead just rely fully on the browser to decide what's focusable.

Can revisit this later if it ends up being a problem.